src/cpu.rs

```diff
@@ -17,23 +20,222 @@ pub struct CPU {
     pub r7: u16,
     pub pc: u16,
     pub cond: u16,
+    pub memory: Memory,
 }
 
 impl CPU {
     pub fn new() -> Self {
         Self {
+            r0: 0,
+            r1: 0,
+            r2: 0,
+            r3: 0,
+            r4: 0,
+            r5: 0,
+            r6: 0,
+            r7: 0,
             pc: 0x3000,
-            ..Default::default()
+            cond: 0,
+            memory: Memory::new(),
         }
     }
 
-    pub fn fetch_instruction(&mut self, memory: &Memory) -> Option<u16> {
-        let instruction = memory.read(self.pc.into())?;
+    pub fn fetch_instruction(&mut self) -> Option<u16> {
+        let instruction = self.memory.read(self.pc.into())?;
         self.pc = self.pc.checked_add(1)?;
 
         let op = Opcode::from(instruction);
-        println!("{:#?}", op);
 
         Some(instruction)
     }
+
+    pub fn execute(&mut self, opcode: Opcode) -> Result<(), CPUErrors> {
+        match opcode {
+            Opcode::OP_ADD_SR { dr, sr1, sr2 } => {
+                let src_register = self.get_register_value(sr1)?;
+                let rhs_register = self.get_register_value(sr2)?;
+                let sum = src_register.wrapping_add(rhs_register);
+                self.update_register(dr, sum)
+                    .map_err(|_| CPUErrors::Execute)?;
+
+                self.update_flag(dr)?;
```

Comment:
Maybe you could move the `update_flag` function call inside the `update_register` function. I believe every time a register is updated the flags need to be updated. 

Mainly to avoid possibly forgetting to call the `update_flag` function AFTER `update_register`. Just a suggestion. 

If you do decide to change it, be mindful about variable placing, the borrow checker might become your enemy 😠 🦀